### ISSUE TITLE
Branch `push` build being mis-attributed to master causing duplicate statuses/deploys

### ISSUE THREAD

**Issue by edmorley** [travis-ci/travis-ci#5008]
tl;dr: I'm seeing multiple Heroku deployments for the same master revision, since GitHub is sending seemingly duplicate `status` event payloads from the Travis builds - because the branch is mis-reported as `master` not `my-feature-branch` in the `branches` key of the payload.

Context:

A common workflow for me:
1) Rebase a pull-request's branch locally/squash fixups
2) Force push to the remote branch on GitHub
3) Very soon after, push that branch to the remote master
...so that GitHub detects my push to master as being the same contents as the PR, and automatically closes the PR. (Rather than me having to add the "Manually merged as revision X" and press the close button, leaving the PR as status red not purple).

With this workflow, three builds are typically running at the same time:
- `continuous-integration/travis-ci/push` for `my-feature-branch`
- `continuous-integration/travis-ci/pr` for `my-feature-branch`
- `continuous-integration/travis-ci/push` for `master`

This results in three GitHub `status` events being sent to webhooks when those builds complete, around about the same time. Heroku needs to only listen to events relating to `master` and also needs to handle their being `push` vs `pr` build statuses. 

However Heroku is not able to do this, since the two events for the 'push' builds both report as being for branch `master` (see the `branches` key in the payload below), when the first payload here is actually for the `spidermonkey-arm64-sim` branch according to the Travis job page (follow the URL under the `target_url` key):

X-GitHub-Delivery: ace39f00-79dd-11e5-8da5-50aa9044a2f7

``` json
{
  "id": 342433058,
  "sha": "afcb9d04eeab6b518a9557dffee1f9989c6f0036",
  "name": "mozilla/treeherder",
  "target_url": "https://travis-ci.org/mozilla/treeherder/builds/87136641",
  "context": "continuous-integration/travis-ci/push",
  "description": "The Travis CI build passed",
  "state": "success",
  "commit": {
    "sha": "afcb9d04eeab6b518a9557dffee1f9989c6f0036",
    "commit": {
      "author": {
        "name": "Ed Morley",
        "date": "2015-10-23T00:02:01Z"
      },
      "committer": {
        "name": "Ed Morley",
        "date": "2015-10-23T23:01:14Z"
      },
      "message": "Bug 1210475 - Add support for SpiderMonkey ARM64 Simulator Build",
      "tree": {
        "sha": "ab65963e13cadc4e9b7ed0370bd9b21c5eb5aa09",
        "url": "https://api.github.com/repos/mozilla/treeherder/git/trees/ab65963e13cadc4e9b7ed0370bd9b21c5eb5aa09"
      },
      "url": "https://api.github.com/repos/mozilla/treeherder/git/commits/afcb9d04eeab6b518a9557dffee1f9989c6f0036",
      "comment_count": 0
    },
    "url": "https://api.github.com/repos/mozilla/treeherder/commits/afcb9d04eeab6b518a9557dffee1f9989c6f0036",
    "parents": [
      {
        "sha": "92693f5be89930cc354e309b1baab2ce58ae156a",
        "url": "https://api.github.com/repos/mozilla/treeherder/commits/92693f5be89930cc354e309b1baab2ce58ae156a",
        "html_url": "https://github.com/mozilla/treeherder/commit/92693f5be89930cc354e309b1baab2ce58ae156a"
      }
    ]
  },
  "branches": [
    {
      "name": "master",
      "commit": {
        "sha": "afcb9d04eeab6b518a9557dffee1f9989c6f0036",
        "url": "https://api.github.com/repos/mozilla/treeherder/commits/afcb9d04eeab6b518a9557dffee1f9989c6f0036"
      }
    }
  ],
  "created_at": "2015-10-23T23:27:50Z",
  "updated_at": "2015-10-23T23:27:50Z"
}
```

X-GitHub-Delivery: 4612b580-79de-11e5-9332-8753a1948893

``` json
{
  "id": 342435803,
  "sha": "afcb9d04eeab6b518a9557dffee1f9989c6f0036",
  "name": "mozilla/treeherder",
  "target_url": "https://travis-ci.org/mozilla/treeherder/builds/87137183",
  "context": "continuous-integration/travis-ci/push",
  "description": "The Travis CI build passed",
  "state": "success",
  "commit": {
    "sha": "afcb9d04eeab6b518a9557dffee1f9989c6f0036",
    "commit": {
      "author": {
        "name": "Ed Morley",
        "date": "2015-10-23T00:02:01Z"
      },
      "committer": {
        "name": "Ed Morley",
        "date": "2015-10-23T23:01:14Z"
      },
      "message": "Bug 1210475 - Add support for SpiderMonkey ARM64 Simulator Build",
      "tree": {
        "sha": "ab65963e13cadc4e9b7ed0370bd9b21c5eb5aa09",
        "url": "https://api.github.com/repos/mozilla/treeherder/git/trees/ab65963e13cadc4e9b7ed0370bd9b21c5eb5aa09"
      },
      "url": "https://api.github.com/repos/mozilla/treeherder/git/commits/afcb9d04eeab6b518a9557dffee1f9989c6f0036",
      "comment_count": 0
    },
    "url": "https://api.github.com/repos/mozilla/treeherder/commits/afcb9d04eeab6b518a9557dffee1f9989c6f0036",
    "parents": [
      {
        "sha": "92693f5be89930cc354e309b1baab2ce58ae156a",
        "url": "https://api.github.com/repos/mozilla/treeherder/commits/92693f5be89930cc354e309b1baab2ce58ae156a",
        "html_url": "https://github.com/mozilla/treeherder/commit/92693f5be89930cc354e309b1baab2ce58ae156a"
      }
    ]
  },
  "branches": [
    {
      "name": "master",
      "commit": {
        "sha": "afcb9d04eeab6b518a9557dffee1f9989c6f0036",
        "url": "https://api.github.com/repos/mozilla/treeherder/commits/afcb9d04eeab6b518a9557dffee1f9989c6f0036"
      }
    }
  ],
  "created_at": "2015-10-23T23:32:07Z",
  "updated_at": "2015-10-23T23:32:07Z"
}
```

ie: For the first payload above I would have expected `branches.name` to be `spidermonkey-arm64-sim` not `master`.

Is Travis mis-reporting the branch since by the time the build has completed, the PR has already been merged?

Many thanks :-)

**Comment by edmorley** [comment]
Hi - is there any news on this?
Thanks :-)

**Comment by edmorley** [comment]
This is still happening - today someone pushed to the `master` branch, then whilst that push's Travis run was still running they pushed the same SHA to the `production` branch, which triggered it's own Travis push job.

There were then two GitHub events delivered - one for each of `master` and `production` branches, however the contents were virtually identical for both, and there is no way to tell which "belonged" to each branch without visiting the Travis build URL, since the payload's `branches` listed all branches:

```json
  "branches": [
    {
      "name": "master",
      "commit": {
        "sha": "9332b6a770c6c0e45a99cdb0c3dd550922fe7cc9",
        "url": "https://api.github.com/repos/mozilla/treeherder/commits/9332b6a770c6c0e45a99cdb0c3dd550922fe7cc9"
      }
    },
    {
      "name": "production",
      "commit": {
        "sha": "9332b6a770c6c0e45a99cdb0c3dd550922fe7cc9",
        "url": "https://api.github.com/repos/mozilla/treeherder/commits/9332b6a770c6c0e45a99cdb0c3dd550922fe7cc9"
      }
    }
  ],
```

This caused duplicated Heroku deploys in multiple environments.

Full diff of the two payloads:
https://emorley.pastebin.mozilla.org/8984072

Looking at the GitHub API docs, it seems like this may actually be a deficiency in the API, since it just uses the SHA and no other specifier:
https://developer.github.com/v3/repos/statuses/

**Comment by DrTorte** [comment]
Hey there,

Don't have any updates on this one but I'm pinging internally to see if we can get some light shed on this. Thanks for all the details and sorry for the automatic closure!

**Comment by cotsog** [comment]
@edmorley What would be helpful for us is to get links/URLs to recent builds where this is still happening. Could you either copy/paste them here or send them over to `support [at] travis-ci [dot] com` ?
